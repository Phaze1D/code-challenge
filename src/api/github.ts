import axios from 'axios';
import { SearchReposParams, GetRepoParams } from 'types';

const githubAxios = axios.create({
  baseURL: 'https://api.github.com',
  headers: {
    common: {
      Accept: 'application/vnd.github.v3+json'
    }
  }
});
export const __githubAxios = githubAxios;


export const setToken = (token: string) => {
  githubAxios.defaults.headers.common.Authorization = `token ${token}`;
};


export const searchRepos = (params: SearchReposParams, cancelToken?) => {
  return githubAxios.get('/search/repositories', {params, cancelToken});
};

export const getRepo = (params: GetRepoParams, cancelToken?) => {
  const {owner, name} = params;
  return githubAxios.get(`/repos/${owner}/${name}`, {cancelToken});
};
