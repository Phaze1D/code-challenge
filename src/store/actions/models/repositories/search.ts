import { omit } from 'lodash';
import { AxiosResponse } from 'axios';
import { github } from 'api';
import { SearchReposParams } from 'types';
import { REPLACE_MANY, UPSERT_MANY } from 'store/actions/types';
import { handleRequestFail } from 'store/actions/helpers';


export const search = (params: SearchReposParams, replace = false) => {
  return dispatch => {
    return github.searchRepos(params)
      .then(res => dispatch(action(res, replace)))
      .catch(error => dispatch(handleRequestFail(error)));
  };
};

const action = (res: AxiosResponse, replace = false) => {
  const {data} = res;

  const models = data.items.map(repo => ({
    type: 'Repository',
    id: repo.id,
    data: repo
  }));

  models.push({
    type: 'Search',
    id: 'searchID',
    data: omit(data, 'items')
  });

  return {
    type: replace ? REPLACE_MANY : UPSERT_MANY,
    payload: {models}
  };
};
