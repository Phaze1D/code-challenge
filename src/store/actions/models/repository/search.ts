import { omit } from 'lodash';
import { AxiosResponse } from 'axios';
import { github } from 'api';
import { SearchReposParams } from 'types';
import { REPLACE_MANY, UPSERT_MANY, UPSERT, REPLACE } from 'store/actions/types';
import { handleRequestFail } from 'store/actions/helpers';


export const search = (params: SearchReposParams, replace = false) => {
  return (dispatch) => {
    dispatch(paginationAction(params)); // TODO: Need to add test for this
    return github.searchRepos(params)
      .then(res => dispatch(action(res, replace)))
      .catch(error => {
        dispatch({
          type: REPLACE,
          payload: {
            model: {
              type: 'Search',
              id: '0',
              data: {
                loading: false
              }
            }
          }
        });
        return dispatch(handleRequestFail(error));
      });
  };
};

const action = (res: AxiosResponse, replace = false) => {
  const {data, config} = res;
  const models = data.items.map(repo => ({
    type: 'Repository',
    id: repo.id,
    data: repo
  }));

  models.push({
    type: 'Search',
    id: '0',
    data: {
      ...omit(data, 'items'),
      params: config.params,
      loading: false
    }
  });

  return {
    type: replace ? REPLACE_MANY : UPSERT_MANY,
    payload: {models}
  };
};


const paginationAction = (params: SearchReposParams) => {
  return {
    type: UPSERT,
    payload: {
      model: {
        type: 'Search',
        id: '0',
        data: {
          params,
          loading: true
        }
      }
    }
  };
};
