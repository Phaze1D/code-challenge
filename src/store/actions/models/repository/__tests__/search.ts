import MockAdapter from 'axios-mock-adapter';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { githubAxios } from 'api/github';
import { search } from '../search';

const mockStore = configureMockStore([thunk]);

describe('Actions - Search for Repos', () => {
  const mock = new MockAdapter(githubAxios);

  it('Should return correctly transform data', async () => {
    mock.onGet('/search/repositories').reply(200, fixtures);
    const store = mockStore();
    const results = await store.dispatch(search({q: 'react'}));
    expect(results).toMatchSnapshot();
  });

  it('Should catch error and return Error model', async () => {
    mock.onGet('/search/repositories').reply(400, fixtures);
    const store = mockStore();
    const results = await store.dispatch(search({q: 'react'}));
    expect(results).toMatchSnapshot();
  });
});


const fixtures = {
  total_count: 1157653,
  incomplete_results: false,
  items: [
    {
      id: 10270250,
      name: 'react',
      full_name: 'facebook/react'
    },
    {
      id: 10270256,
      name: 'react-hack',
      full_name: 'react-hack'
    }
  ]
};
