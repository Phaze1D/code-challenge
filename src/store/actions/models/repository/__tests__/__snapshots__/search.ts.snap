// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions - Search for Repos Should catch error and return Error model 1`] = `
Object {
  "payload": Object {
    "models": Array [
      Object {
        "data": Object {
          "alertMessage": "Something seems to have gone wrong",
          "error": [Error: Request failed with status code 400],
        },
        "id": "0",
        "type": "Error",
      },
      Object {
        "data": Object {
          "loading": false,
        },
        "id": "0",
        "type": "Search",
      },
    ],
  },
  "type": "REPLACE_MANY",
}
`;

exports[`Actions - Search for Repos Should return correctly transform data 1`] = `
Object {
  "payload": Object {
    "models": Array [
      Object {
        "data": Object {
          "full_name": "facebook/react",
          "id": 10270250,
          "name": "react",
        },
        "id": 10270250,
        "type": "Repository",
      },
      Object {
        "data": Object {
          "full_name": "react-hack",
          "id": 10270256,
          "name": "react-hack",
        },
        "id": 10270256,
        "type": "Repository",
      },
      Object {
        "data": Object {
          "incomplete_results": false,
          "loading": false,
          "params": Object {
            "q": "react",
          },
          "total_count": 1157653,
        },
        "id": "0",
        "type": "Search",
      },
    ],
  },
  "type": "UPSERT_MANY",
}
`;
